src/importer.rs

```diff
@@ -0,0 +1,284 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::fs;
+use std::io::ErrorKind;
+
+use crc32fast;
+use tobj;
+
+use crate::viewport_renderer::{Index, Vertex};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ImporterError {
+    FileNotFound,
+    PermissionDenied,
+    InvalidStructure,
+    Other,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Model {
+    pub name: String,
+    pub vertices: Vec<Vertex>,
+    pub indices: Vec<Index>,
+}
+
+/// `Importer` takes care of importing of obj files and caching of their
+/// internal representations. It holds paths to files, their checksums and
+/// parsed obj files.
+#[derive(Debug, Default)]
+pub struct Importer {
+    pub path_checksums: HashMap<String, u32>,
```

Comment:
Hmmmmm, aren't just 2 fields sufficient?

```rust
checksum_models: HashMap<u32, Vec<Model>>,
path_checksums: HashMap<String, u32>,
```

"Duplicating" would just be adding a entry to `path_checksums`, no need to clone the actual model. This is of course assuming that two checksums for different models aren't the same, but you are already assuming that.

Some logic would change under  this model, e.g. if a file changes, you add the newly parsed model under a new entry to `checksum_models` and rewire the entry in `path_checksums` to it. Once no `path_checksums` point to an entry in `checksum_models`, you can garbage collect it.

I am not sure if this is overall a better approach, but it doesn't duplicate/clone data nearly as much, so it should be more 🍀

Reply:
You got me at 🍀 . Will look into this.